toolz/utils.py

```diff
@@ -6,4 +6,11 @@ def raises(err, lamda):
         return True
 
 
-no_default = '__no__default__'
+@object.__new__
+class no_default(object):
+    def __new__(self):
+        raise TypeError("cannot create 'no_default' instances")
+
+    def __str__(self):
+        return '<object no_default>'
+    __repr__ = __str__
```

Comment:
My main concern with the original implementation is that `__no_default__` is a valid element of the domain for the default value. If you use a sentinal object and do pointer equality on it, the only way to get that value is by explicitly importing it which means you know what you are doing. The sentinal object effectivly removes the chance of ignoring a default that you actually wanted.

The instance of `no_default` here is to address @eriknw's concern that we lost information when debugging.